### ISSUE TITLE
match raises TypeError with elemMatch if field matched does not exist

### ISSUE THREAD

**Issue by jo-tham** [kapouille/mongoquery#12]
**Actual Behaviour**

Using ElemMatch operator against a field which isn't iterable raises `TypeError`

Haven't verified, but probably happens with other operators that are natural to use with arrays

**Desired Behaviour**

Return no match

Perhaps QueryError would be suitable, since ElemMatch shouldn't be used against fields that aren't iterable. But I think the convention is not to have any expectations about the types of the fields

Glad to make a PR here if maintainer can advise what the behavior should be

**Minimal working example**

```
from mongoquery import Query, QueryError

record = {}

subscription = Query({
    "ratios": {
        "$elemMatch": {
            "$gte": {
                "confidence": 0.5
            }
        }
    }
})
```

Traceback

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-630f64a0db6f> in <module>()
----> 1 subscription.match(record)

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in match(self, entry)
     25 
     26     def match(self, entry):
---> 27         return self._match(self._definition, entry)
     28 
     29     def _match(self, condition, entry):

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in _match(self, condition, entry)
     31             return all(
     32                 self._process_condition(sub_operator, sub_condition, entry)
---> 33                 for sub_operator, sub_condition in condition.items()
     34             )
     35         else:

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in <genexpr>((sub_operator, sub_condition))
     31             return all(
     32                 self._process_condition(sub_operator, sub_condition, entry)
---> 33                 for sub_operator, sub_condition in condition.items()
     34             )
     35         else:

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in _process_condition(self, operator, condition, entry)
     98                 return False
     99             extracted_data = entry[operator]
--> 100         return self._match(condition, extracted_data)
    101 
    102     ##################

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in _match(self, condition, entry)
     31             return all(
     32                 self._process_condition(sub_operator, sub_condition, entry)
---> 33                 for sub_operator, sub_condition in condition.items()
     34             )
     35         else:

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in <genexpr>((sub_operator, sub_condition))
     31             return all(
     32                 self._process_condition(sub_operator, sub_condition, entry)
---> 33                 for sub_operator, sub_condition in condition.items()
     34             )
     35         else:

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in _process_condition(self, operator, condition, entry)
     88             if operator.startswith("$"):
     89                 try:
---> 90                     return getattr(self, "_" + operator[1:])(condition, entry)
     91                 except AttributeError:
     92                     raise QueryError(

/home/joth/projects/mongoquery-speed/venv/lib/python2.7/site-packages/mongoquery/__init__.pyc in _elemMatch(self, condition, entry)
    293                 for sub_operator, sub_condition in condition.items()
    294             )
--> 295             for element in entry
    296         )
    297 

TypeError: '_Undefined' object is not iterable

```

**Comment by kerfab** [comment]
Do you think a control such as:

    def _elemMatch(self, condition, entry):
        if not isinstance(entry, Sequence):
            return False

could do the job?

**Comment by jo-tham** [comment]
Yep something like that seems reasonable @kerfab

**Comment by kerfab** [comment]
I will see that this gets fixed on the current [PR #11](https://github.com/kapouille/mongoquery/pull/11)

**Comment by kerfab** [comment]
Fixed in 1.3.0
